### PR TITLE
Improved docs for NDSL examples

### PR DESCRIPTION
**Pull request by romanc**
**Description**

Tutorials are in important step in on-boarding new library users. While working through the examples, I noticed a couple of typos and broken links. In the bigger picture, we figured that examples should use the boilerplate code that the middleware already exposes in `ndsl.boilerplate`. I've also added a README file with a little bit of context and a quickstart guide to get people up and running with minimal overhead. And we agreed to remove the output currently stored in the Jupyter notebooks. This was one in the first commit such that we can see the actual changes when looking at the other commits.

In summary

- Removed all output from the Jupyter notebooks
- Moved `plot_filed_at_kN` from `examples/NDSL/basic_boilerplate.py`  to ` ndsl/boilerplate.py`. This adds `matplotlib` to the list of dependnecies.
- Refactored Jupyter notebooks get all boilerplate code from `ndsl.boilerplate`.
- Removed `examples/NDSL/*_boilerplate.py`
- Added `examples/NDSL/README.md` with a bit of context.
- Fixed typos and broken links in Jupyter notebooks.

It is recommended to exclude the first commit when looking at the diff, e.g. [this view](https://github.com/NOAA-GFDL/NDSL/pull/63/files/ffd042912b6a20c1db240433d144863e056329fc..7fda6d42f05fe92561f6e8bc409f762669160f2c) is much nicer to look at.

Regarding the previous addition of a `.tool-versions` file a python version manager: With the install instructions as simple as `pip install ndsl[demos]`, I don't think we need this complexity anymore. The middleware's README already specifies the need for python 3.11. How developers ensure this should be up to them.

Fixes: N/A

**How Has This Been Tested?**

I ran all the notebooks locally against a local version of NDSL.

**Checklist:**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] New check tests, if applicable, are included

### CODE REVIEW

File: examples/NDSL/requirements.txt
Comment (FlorianDeconinck) [inline examples/NDSL/requirements.txt:1]: Can we fold this into the `setup.py` directly under a `ndsl[demos]` flavor or similar?

File: examples/NDSL/basic_boilerplate.py
Comment (FlorianDeconinck) [inline examples/NDSL/basic_boilerplate.py:1]: This file needs to be refactored. `ndsl` carries now a series of boilerplate codes as a first citizen (see `ndsl.boilerplate`)

We can keep the plotting there, even if we should move those as features of `Quantity` as a tool

**Review by FlorianDeconinck — COMMENTED** [review]
I am for removing all results here and encoding them in the markdown and/or rendering the plots.

File: examples/NDSL/requirements.txt
Comment (romanc) [reply examples/NDSL/requirements.txt:1]: I like that idea :+1:

File: examples/NDSL/basic_boilerplate.py
Comment (romanc) [reply examples/NDSL/basic_boilerplate.py:1]: Sounds good - let me check what is there already and move the rest.

File: examples/NDSL/02_NDSL_basics.ipynb [hunk 1]
```diff
@@ -25,16 +25,15 @@
...
+    "from ndsl.boilerplate import get_factories_single_tile_numpy, plot_field_at_kN\n",
     "\n",
     "nx = 6\n",
     "ny = 6\n",
     "nz = 1\n",
     "nhalo = 1\n",
-    "backend=\"numpy\"\n",
     "\n",
-    "stencil_factory: StencilFactory = get_one_tile_factory(nx, ny, nz, nhalo, backend)"
+    "stencil_factory, _ = get_factories_single_tile_numpy(nx, ny, nz, nhalo)"
```
Comment (romanc) [inline examples/NDSL/02_NDSL_basics.ipynb:36]: Unsure about python syntax here (about the `_`): is this common in python or is there another convention for unused destructured arguments?

File: examples/NDSL/02_NDSL_basics.ipynb [hunk 2]
```diff
@@ -125,12 +124,11 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "## Change this to Quantity\n",
-    "\n",
```
Comment (romanc) [inline examples/NDSL/02_NDSL_basics.ipynb:166]: The code below is using `Quantity` objects. Seems like a lost todo. Or is there more refactoring needed here?

File: setup.py [hunk 3]
```diff
@@ -29,6 +34,7 @@ def local_pkg(name: str, relative_path: str) -> str:
     "h5netcdf",  # for xarray
     "dask",  # for xarray
     "numpy==1.26.4",
+    "matplotlib",  # for plotting in boilerplate
```
Comment (romanc) [inline setup.py:37]: `ndsl/boilerplate.py`  is used in a bunch of places, e.g. in unit tests. I thus couldn't add `matplotlib` only for the `[demos]` flavor.

**Review by romanc — COMMENTED** [review]
Some notes/questions inline

File: examples/NDSL/02_NDSL_basics.ipynb (hunk 1 above)
Comment (FlorianDeconinck) [reply examples/NDSL/02_NDSL_basics.ipynb:36]: I think that still the convention, or at least I am unaware of something better

File: setup.py (hunk 3 above)
Comment (FlorianDeconinck) [reply setup.py:37]: Not great, but we were heading there anyway

**Review by FlorianDeconinck — CHANGES_REQUESTED** [review]

**Comment by romanc** [comment]
Sorry, for yet another commit. Just realized that one of the notebooks was marked as executable

image

**Review by FlorianDeconinck — APPROVED** [review]
Looking good, merging this

**Pull request merged**